### PR TITLE
fix(fxconfig): cleaning up stale subscriber on canceled/timeout Subscribe paths

### LINKED ISSUE #109: fxconfig: Subscribe leaks stale subscriber on canceled context
While going through the notification flow, I noticed a subtle issue around how subscribers are handled when a context is already canceled. Right now, the subscriber entry is added before checking whether the context is valid. If the context is canceled, the function exits early, but the subscriber is never cleaned up.

This can leave behind a stale subscriber for a given txID, which can interfere with future subscriptions. In particular, a later call with a valid context may incorrectly assume that a subscription already exists and skip opening a proper upstream request, leading to hangs or missed notifications.

A simple way to reproduce this is to first call Subscribe with a canceled context and then call it again with an active context for the same txID. The second call may not behave correctly because of the leftover state.

It would be safer to either insert the subscriber only after successfully enqueueing the request, or explicitly roll back the subscriber entry if the context is canceled or an error occurs.

### PR DESCRIPTION
**Pull request by rootp1**
## Problem

`NotificationClient.Subscribe()` appended a subscriber before validating context state. If the context was canceled before queueing the upstream request, the method returned with an error but left stale subscriber state in memory for the same txID.

## Root cause

In `Subscribe()` the map mutation happened too early:

1. Create receiver channel
2. Append subscriber to `subscribers[txID]`
3. Later fail on canceled/timed-out context before successful queue send
4. Stale subscriber remained

## Implemented fix

Updated `Subscribe()` to keep single-upstream-subscription behavior and clean up stale state on failed enqueue:

- Keep subscriber registration under lock
- If this is not the first subscriber for a txID, return immediately (reuse existing upstream subscription)
- For first subscriber path, if context is canceled before enqueue, delete `subscribers[txID]`
- If context is canceled while waiting to enqueue, delete `subscribers[txID]`

This guarantees the map is rolled back when request enqueue does not happen.

### CODE REVIEW

**Comment by rootp1** [comment]
@mbrandenburger, Please have a look, whenever you have a chance to. Thankyou

File: tools/fxconfig/internal/client/notifications.go [hunk 1]
```diff
@@ -114,13 +113,15 @@ func (n *NotificationClient) Subscribe(ctx context.Context, txID string) (chan i
...
 	case <-ctx.Done():
+		delete(n.subscribers, txID)
 		return nil, ctx.Err()
 	default:
 	}
 
 	// try to push to request queue
 	select {
 	case <-ctx.Done():
+		delete(n.subscribers, txID)
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/client/notifications.go]: this deletion comes with the problem that we just remove all subscribers, if there are any and not just the one we are adding here.

**Review by mbrandenburger — CHANGES_REQUESTED** [review]
Thank you @rootp1 for creating this PR. I think this is good start. 

Please have a look at https://github.com/hyperledger/fabric-x/pull/123. Maybe we can cherry pick the changes to `notifications.go` and add you tests. WDYT?

File: tools/fxconfig/internal/client/notifications.go (hunk 1 above)
Comment (rootp1) [reply tools/fxconfig/internal/client/notifications.go]: Thanks for calling this out @mbrandenburger, I’ve addressed it in the latest update.

In notifications.go, I replaced the broad delete(n.subscribers, txID) rollback with targeted cleanup that removes only the subscriber channel created by the current Subscribe call. 

If other subscribers already exist for the same txID, they are preserved.

I also added a regression test to verify that canceling one blocked subscribe path does not remove existing subscribers for the same txID.

Please take another look when you have a moment.

**Comment by rootp1** [comment]
> Thank you @rootp1 for creating this PR. I think this is good start.
> 
> Please have a look at #123. Maybe we can cherry pick the changes to `notifications.go` and add you tests. WDYT?

As, discussed I am up with the changes suggested by you @mbrandenburger. Please have a look whenever you have a chance to

File: tools/fxconfig/internal/client/notifications.go (hunk 1 above)
Comment (rootp1) [reply tools/fxconfig/internal/client/notifications.go]: also, my regression test verifies confirms that when one blocked Subscribe call is canceled, only that call’s subscriber is removed, and any other existing subscriber for the same txID remains registered.

**Comment by cynox-66** [comment]
Hey @rootp1 
This looks much better now — the targeted cleanup makes sense 👍

One thing I was thinking about while reading this: `removeSubscriber` and the dispatcher in `listen()` both end up touching the same map. The locking + guard makes it safe, but it might be worth a quick comment there just to make that interaction obvious for anyone revisiting this later.

Also nice catch on the regression test — that’s a subtle edge case.

**Comment by rootp1** [comment]
> Hey @rootp1 This looks much better now — the targeted cleanup makes sense 👍
> 
> One thing I was thinking about while reading this: `removeSubscriber` and the dispatcher in `listen()` both end up touching the same map. The locking + guard makes it safe, but it might be worth a quick comment there just to make that interaction obvious for anyone revisiting this later.
> 
> Also nice catch on the regression test — that’s a subtle edge case.

Thankyou.

File: tools/fxconfig/internal/client/notifications.go [hunk 2]
```diff
@@ -114,20 +114,51 @@ func (n *NotificationClient) Subscribe(ctx context.Context, txID string) (chan i
...
 	case <-ctx.Done():
+		n.removeSubscriber(txID, receiverCh)
 		return nil, ctx.Err()
 	case n.requestQueue <- req:
 	}
 
 	return receiverCh, nil
 }
 
+func (n *NotificationClient) removeSubscriber(txID string, receiverCh chan int) {
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/client/notifications.go]: I liked the the small inline rollback function @cynox-66 used in his PR. In particular, because it's only used within `Subscribe()`. Do you see any advantage to promote `removeSubscriber` to a method of `NotificationClient`?

```go
rollback := func() {
....    
}
```

File: tools/fxconfig/internal/client/notifications.go [hunk 3]
```diff
@@ -91,17 +91,17 @@ func (n *NotificationClient) Subscribe(ctx context.Context, txID string) (chan i
...
 
 	subscribers := n.subscribers[txID]
 	n.subscribers[txID] = append(subscribers, receiverCh)
 
 	if len(subscribers) > 0 {
 		// we already have an active subscription for this txID
+		n.subscribersMu.Unlock()
 		return receiverCh, nil
 	}
+	n.subscribersMu.Unlock()
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/client/notifications.go]: replacing `defer Unlock()` with two explicit `Unlock()` seem to work but we could mess this code up later easily. Do you think something like the following would work?

```go
  isFirst := func() bool {
      n.subscribersMu.Lock()
      defer n.subscribersMu.Unlock()
      existing := n.subscribers[txID]
      // update the subscribers 
      n.subscribers[txID] = append(existing, receiverCh)
      return len(existing) == 0
  }()
  if !isFirst {
      return receiverCh, nil
  }
```

**Review by mbrandenburger — CHANGES_REQUESTED** [review]
Thank you @rootp1 for working on this PR! Good progress. I left a few comments below.

File: tools/fxconfig/internal/client/notifications.go (hunk 2 above)
Comment (rootp1) [reply tools/fxconfig/internal/client/notifications.go]: I agree with keeping this local to Subscribe(). 
I refactored it into an inline rollback closure since it’s only used on the subscribe path and doesn’t add reusable behavior as a client-wide method. This also keeps subscription mutation and rollback logic co-located.

one possible advantage of a NotificationClient method is reuse if the same cleanup is needed from multiple call sites, but right now keeping it local in Subscribe() is clearer and safer because it captures only function-local context and avoids widening the public/internal API surface.

File: tools/fxconfig/internal/client/notifications.go (hunk 3 above)
Comment (rootp1) [reply tools/fxconfig/internal/client/notifications.go]: Thankyou for the suggestion @mbrandenburger, I switched to exactly that pattern via an inline isFirst function (Lock + defer Unlock) and early return when it’s not the first subscriber. 

That removes the double explicit unlock path and makes the locking logic safer to maintain.

**Comment by rootp1** [comment]
@mbrandenburger, please have a look whenever you have a chance to.

**Comment by mbrandenburger** [comment]
@cynox-66 could you also have another look please?

**Review by mbrandenburger — APPROVED** [review]

**Comment by cynox-66** [comment]
Hey @mbrandenburger  — took another look after the latest changes.

The `isFirst` pattern + inline rollback makes the flow much cleaner now, and the targeted cleanup fixes the original issue without affecting existing subscribers.
The regression test also covers the tricky cancellation case well.

Looks good from my side 👍

**Comment by mbrandenburger** [comment]
Thank you @rootp1 @cynox-66 great work!

**Comment by rootp1** [comment]
> Thank you @rootp1 @cynox-66 great work!

Thankyou @mbrandenburger, for the appreciation

**Comment by cynox-66** [comment]
Thank you @mbrandenburger .
